src/utils/GroupCallUtils.ts

```diff
@@ -0,0 +1,150 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { EventTimeline, MatrixClient, MatrixEvent, RoomState } from "matrix-js-sdk/src/matrix";
+import { deepCopy } from "matrix-js-sdk/src/utils";
+
+export const STUCK_DEVICE_TIMEOUT_MS = 1000 * 60 * 60; // 1 hour
+
+const CALL_STATE_EVENT_TYPE = "org.matrix.msc3401.call";
+const CALL_MEMBER_STATE_EVENT_TYPE = "org.matrix.msc3401.call.member";
+const CALL_STATE_EVENT_TERMINATED = "m.terminated";
+
+interface MDevice {
+    ["m.device_id"]: string;
+}
+
+interface MCall {
+    ["m.call_id"]: string;
+    ["m.devices"]: Array<MDevice>;
+}
+
+interface MCallMemberContent {
+    ["m.expires_ts"]: number;
+    ["m.calls"]: Array<MCall>;
+}
+
+const getRoomState = (client: MatrixClient, roomId: string): RoomState => {
+    return client.getRoom(roomId)
+        ?.getLiveTimeline()
+        ?.getState?.(EventTimeline.FORWARDS);
+};
+
+/**
+ * Finds the latest, non-terminated call state event.
+ */
+export const getGroupCall = (client: MatrixClient, roomId: string): MatrixEvent => {
+    return getRoomState(client, roomId)
+        ?.getStateEvents(CALL_STATE_EVENT_TYPE)
+        .sort((a: MatrixEvent, b: MatrixEvent) => b.getTs() - a.getTs())
+        .find((event: MatrixEvent) => {
+            return !(CALL_STATE_EVENT_TERMINATED in event.getContent());
+        });
+};
+
+/**
+ * Finds the "m.call.member" events for an "m.call" event.
+ *
+ * @returns {MatrixEvent[]} non-expired "m.call.member" events for the call
+ */
+export const useConnectedMembers = (client: MatrixClient, callEvent: MatrixEvent): MatrixEvent[] => {
+    if (callEvent.getType() !== CALL_STATE_EVENT_TYPE) return [];
+
+    const callId = callEvent.getStateKey();
+    const now = Date.now();
+    return getRoomState(client, callEvent.getRoomId())
+        ?.getStateEvents(CALL_MEMBER_STATE_EVENT_TYPE)
+        .filter((callMemberEvent: MatrixEvent): boolean => {
+            const {
+                ["m.expires_ts"]: expiresTs,
+                ["m.calls"]: calls,
+            } = callMemberEvent.getContent<MCallMemberContent>();
+
+            // state event expired
+            if (expiresTs && expiresTs < now) return false;
+
+            return !!calls?.find((call: MCall) => call["m.call_id"] === callId);
+        }) || [];
+};
+
+/**
+ * Removes a list of devices from a call.
+ * Only works for the current user's devices.
+ */
+const removeDevices = async (client: MatrixClient, callEvent: MatrixEvent, deviceIds: string[]) => {
+    if (callEvent.getType() !== CALL_STATE_EVENT_TYPE) return;
+
+    const roomId = callEvent.getRoomId();
+    const callMemberEvent = getRoomState(client, roomId)
+        ?.getStateEvents(CALL_MEMBER_STATE_EVENT_TYPE, client.getUserId());
+    const callMemberEventContent = callMemberEvent?.getContent<MCallMemberContent>();
```

Comment:
I think it might be nicer to assign `callMemberEventContent["m.call"]` to a var and then copy it. Would that work?

Reply:
Instead of copying the entire content? I really want to prevent unexpected mutations on the events. That's why I copying all the content to completely separate it.

Reply:
But does the function mutate anything other than calls? I might missed it

Reply:
It does eventually change the expire timestamp: https://github.com/matrix-org/matrix-react-sdk/pull/9131/files#diff-af078f072ed18f9e5672b00e4b51dc223edfc1d0ae2deea25368722b887a12e5R117

My idea here is to only update the devices list and the expire timestamp of the state event while leaving the rest untouched. No matter if it will have additional properties at any time in the future. In my opinion this should be more stable than assembling a new event content. If I am on the wrong track, please let me know :wink:

Reply:
Yeah, makes sense now! Missed that line